packages/dds/tree/src/feature-libraries/modular-schema/discrepancies.ts

```diff
@@ -0,0 +1,384 @@
+ */
+export function getAllowedContentIncompatibilities(
+	view: TreeStoredSchema,
+	stored: TreeStoredSchema,
+): Incompatibility[] {
+	const incompatibilities: Incompatibility[] = [];
+
+	// check root schema discrepancies
+	incompatibilities.push(
+		...trackFieldDiscrepancies(view.rootFieldSchema, stored.rootFieldSchema),
+	);
+
+	// Verify the existence and type of a node schema given its identifier (key), then determine if
+	// an exhaustive search is necessary.
+	const viewNodeKeys = new Set<TreeNodeSchemaIdentifier>();
+	for (const [key, viewNodeSchema] of view.nodeSchema) {
+		viewNodeKeys.add(key);
+
+		if (viewNodeSchema instanceof ObjectNodeStoredSchema) {
+			if (!stored.nodeSchema.has(key)) {
+				incompatibilities.push({
+					identifier: key,
+					mismatch: "nodeKind",
+					view: "object",
+					stored: undefined,
+				});
+			} else {
+				const storedNodeSchema = stored.nodeSchema.get(key);
+				if (storedNodeSchema instanceof MapNodeStoredSchema) {
+					incompatibilities.push({
+						identifier: key,
+						mismatch: "nodeKind",
+						view: "object",
+						stored: "map",
+					} satisfies NodeKindIncompatibility);
+				} else if (storedNodeSchema instanceof LeafNodeStoredSchema) {
+					incompatibilities.push({
+						identifier: key,
+						mismatch: "nodeKind",
+						view: "object",
+						stored: "leaf",
+					} satisfies NodeKindIncompatibility);
+				} else if (storedNodeSchema instanceof ObjectNodeStoredSchema) {
+					const differences = trackObjectNodeDiscrepancies(viewNodeSchema, storedNodeSchema);
+					if (differences.length > 0) {
+						incompatibilities.push({
+							identifier: key,
+							mismatch: "fields",
+							differences,
+						} satisfies NodeFieldsIncompatibility);
+					}
+				} else {
+					throwUnsupportedNodeType(typeof storedNodeSchema);
+				}
+			}
+		} else if (viewNodeSchema instanceof MapNodeStoredSchema) {
+			if (!stored.nodeSchema.has(key)) {
+				incompatibilities.push({
+					identifier: key,
+					mismatch: "nodeKind",
+					view: "map",
+					stored: undefined,
+				} satisfies NodeKindIncompatibility);
+			} else {
+				const storedNodeSchema = stored.nodeSchema.get(key);
+				if (storedNodeSchema instanceof ObjectNodeStoredSchema) {
+					incompatibilities.push({
+						identifier: key,
+						mismatch: "nodeKind",
+						view: "map",
+						stored: "object",
+					} satisfies NodeKindIncompatibility);
+				} else if (storedNodeSchema instanceof LeafNodeStoredSchema) {
+					incompatibilities.push({
+						identifier: key,
+						mismatch: "nodeKind",
+						view: "map",
+						stored: "leaf",
+					} satisfies NodeKindIncompatibility);
+				} else if (storedNodeSchema instanceof MapNodeStoredSchema) {
+					incompatibilities.push(
+						...trackFieldDiscrepancies(
+							viewNodeSchema.mapFields,
+							storedNodeSchema.mapFields,
+							key,
+						),
+					);
+				} else {
+					throwUnsupportedNodeType(typeof storedNodeSchema);
```

Comment:
here and elsewhere you don't want `typeof`... that will always be the string "function". Since whatever this is should be assignable to `TreeNodeStoredSchema`, I think it should be safe to do `storedNodeSchema.constructor.name` here.